src/cli/platform/sgx/tcb.rs

```diff
@@ -0,0 +1,194 @@
+            ))?
+            .write_all(fmspc.as_bytes())
+            .context(format!("writing Intel FMSPC to file {FMSPC_PATH}"))?;
+
+        Ok(ExitCode::SUCCESS)
+    }
+}
+
+fn decode_extension(cert: &Certificate) -> anyhow::Result<String> {
+    const SGX_EXT_OID: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1");
+    const FMSPC_OID: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.4");
+    const FMSPC_OID_LEN: usize = 10;
+
+    if let Some(extensions) = &cert.tbs_certificate.extensions {
+        for extension in extensions.iter() {
+            if extension.extn_id == SGX_EXT_OID {
+                let offset = extension
+                    .extn_value
+                    .windows(FMSPC_OID_LEN)
+                    .position(|window| window == FMSPC_OID.as_bytes());
+                if let Some(offset) = offset {
+                    // Index into the extension bytes at the offset + OID size + 2
+                    // The extra +2 gets us past the ASN.1 header for these bytes.
+                    let fmspc = &extension.extn_value
+                        [offset + FMSPC_OID_LEN + 2..offset + FMSPC_OID_LEN + 6 + 2];
+                    return Ok(hex::encode_upper(fmspc));
+                }
+            }
+        }
+    }
+    Err(anyhow::Error::msg(
+        "failed to parse SGX certificate extensions, FMSPC not found.",
+    ))
+}
+
+fn fetch_file(url: &str) -> anyhow::Result<(Vec<u8>, Option<String>)> {
+    let response = ureq::get(url).call().context(format!("retrieving {url}"))?;
+
+    let tcb_cert_chain = response.header("tcb-info-issuer-chain").map(String::from);
+
+    let mut reader = response.into_reader();
+
+    let mut bytes = vec![];
+    reader
+        .read_to_end(&mut bytes)
+        .context("reading bytes buffer")?;
+    Ok((bytes, tcb_cert_chain))
+}
+
+/// Fetch the TCB report from Intel
+#[derive(Args, Debug)]
+pub struct TcbCache {}
+
+impl TcbCache {
+    pub fn execute(self) -> anyhow::Result<ExitCode> {
+        if !Path::new(FMSPC_PATH).exists() {
+            bail!("Cannot read {FMSPC_PATH}, please run `enarx platform sgx cache-pck`")
+        }
+
+        let fmspc =
+            std::fs::read_to_string(FMSPC_PATH).context(format!("Failed to read {FMSPC_PATH}"))?;
+
+        let url = format!("{TCB_URL}?fmspc={fmspc}");
+        let (tcb_report, tcb_cert_chain) =
+            fetch_file(&url).context("Failed to fetch Intel TCB report")?;
+
+        if tcb_cert_chain.is_none() {
+            bail!("Did not receive Intel TCB signing certificates in HTTP response header");
+        }
+
+        let mut tcb_cert_chain = tcb_cert_chain.unwrap();
+        // Decode HTTP-friendly characters
+        tcb_cert_chain = tcb_cert_chain
+            .replace("%0A", "\n")
+            .replace("%20", " ")
+            .replace("%2B", "+")
+            .replace("%2F", "/")
+            .replace("%3D", "=");
```

Comment:
I am sure, you want to use something like https://docs.rs/percent-encoding/latest/percent_encoding/fn.percent_decode.html, before getting into trouble not handling all chars.

Reply:
Resolved.